WordPress/src/main/java/org/wordpress/android/ui/domains/DomainRegistrationDetailsFragment.kt

```diff
@@ -197,47 +172,152 @@ class DomainRegistrationDetailsFragment : Fragment() {
+                viewModel.onTosLinkClicked()
             }
         }
+    }
 
-        if (supportedStates == null || supportedStates!!.isEmpty()) {
-            toggleStatesInputFieldVisibility(false)
-        }
-
-        if (savedInstanceState != null) {
-            val isStatesLoadingProgressVisible = savedInstanceState.getBoolean(EXTRA_STATES_FETCH_PROGRESS_VISIBLE)
-            val isFormProgressIndicatorVisible = savedInstanceState.getBoolean(EXTRA_FORM_PROGRESS_INDICATOR_VISIBLE)
-            val isProgressDialogVisible = savedInstanceState.getBoolean(EXTRA_PROGRESS_DIALOG_VISIBLE)
-
-            if (isFormProgressIndicatorVisible) {
-                showFormProgressIndicator()
-            } else {
-                hideFormProgressIndicator()
-            }
+    private fun setupObservers() {
+        viewModel.formProgressIndicatorVisible.observe(this,
+                Observer<Boolean> { isVisible ->
+                    if (isVisible == true) {
+                        showFormProgressIndicator()
+                    } else {
+                        hideFormProgressIndicator()
+                    }
+                })
+
+        viewModel.statesProgressIndicatorVisible.observe(this,
+                Observer<Boolean> { isVisible ->
+                    if (isVisible == true) {
+                        showStateProgress()
+                    } else {
+                        hideStateProgress()
+                    }
+                })
+
+        viewModel.registrationProgressIndicatorVisible.observe(this,
+                Observer<Boolean> { isVisible ->
+                    if (isVisible == true) {
+                        showDomainRegistrationProgressDialog()
+                    } else {
+                        hideDomainRegistrationProgressDialog()
+                    }
+                })
+
+        viewModel.domainRegistrationButtonEnabled.observe(this,
+                Observer<Boolean> { isEnabled ->
+                    register_domain_button.isEnabled = isEnabled == true
+                })
+
+        viewModel.privacyProtectionState.observe(this, object : Observer<Boolean?> {
+            override fun onChanged(privacyEnabled: Boolean?) {
+                if (privacyEnabled != null && privacyEnabled) {
+                    domain_privacy_options_radiogroup.check(R.id.domain_privacy_on_radio_button)
+                } else {
+                    domain_privacy_options_radiogroup.check(R.id.domain_privacy_off_radio_button)
+                }
 
-            if (isStatesLoadingProgressVisible) {
-                showStatesProgressIndicator()
-            } else {
-                hideStatesProgressIndicator()
+                domain_privacy_options_radiogroup.setOnCheckedChangeListener { _, _ ->
+                    viewModel.togglePrivacyProtection(domain_privacy_on_radio_button.isChecked)
+                }
+                viewModel.privacyProtectionState.removeObserver(this)
```

Comment:
Yes, it is absolutely better! And it works really well with 2-way data binding. Without it we have a problem with the data flow loop (?) - we toggle a radio button, listener assigned to it calls ViewModel, which notifies LiveData. Observer reacts to LiveData and toggles the radio button, which triggers a listener and calls ViewModel, etc. :) I replaced `OnCheckedChangeListener` with normal click listener, so it works as intended, but I still remove observer for the `domainContactDetails` data to avoid this (not sure how to avoid doing this).

Reply:
What do you think about this approach:
```
val currentModel = contactFormToDomainContactModel()
                if (currentModel != domainContactModel) {
                    populateContactForm(domainContactModel!!)
                }
```
I'm not 100% sure this will always work but if it doesn't, we just have to make sure the fields are the same (because they will all go through the equals method). It should work because we'll be using the same method to collect the current data and updating the `viewModel.domainContactDetails` field.

Reply:
Ahh, got it! Yes, this will work :)